### ISSUE TITLE
NPM Organization Deploy Tokens

### ISSUE THREAD

**Issue by RobertWHurst** [npm/registry-issue-archive#27]
Me and my team are using the private packages feature NPM offers. The feature is extremely useful, but the setup is rough.

Like most modern dev teams, we use CI for testing and deploying our code. In order to continue using our CI we created a new NPM user and extracted an auth token for that user. Then we created a .npmrc in the root each of our projects using private packages. This .npmrc contains a template reference to an enviroment variable, NPM_TOKEN, in the location where the auth token would normally go. We then set the NPM_TOKEN environment variable to the value of the extracted auth token from the CI user. This seems to be the recommended method as stated in the documentation on npmjs.org.

In addition to the above, we also had to ask each developer to set a NPM_TOKEN environment variable in their shell when working with projects containing private modules. Otherwise npm fails to run within the directory of these projects because the templated .npmrc requires the environment variable to be present. If the user updates their npm token via npm login, they must also remember to copy the new  auth token out of their .npmrc present in their home directory, and update the value of the NPM_TOKEN environment variable in their shell.

This all works, but it's an extremely rough experience. I'd like to suggest some improvements, and I'm happy to help contribute if need be.

Firstly, when one needs to obtain a token for their CI or other automated systems, there should be a location on npmjs.com to create deploy keys for your organization. Having to create a CI user to generate an auth token is a clunky experience.

Lastly, once I have my deploy token I should be able to set it as an environment variable that the NPM cli looks for and uses if found, overriding the .npmrc if present. An alternative to this might be the ability to set the authentication token via a command; npm config set auth-token $NPM_TOKEN. This removes the need for dev to set the NPM_TOKEN environment variable; devs won't need to be concerned with the setup of CI.

So there's my feedback.

Thanks for all of your hard work, NPM is a super bad ass package manager and it's awesome to see you all continuing to improve it. Cheers 🍻

**Comment by enapupe** [comment]
I'd like to add this case where your only option for authenticating is through ENV variables, in this case, setting `.npmrc` is fundamental BUT it forces all developers to have such ENV set, which is a pain

**Comment by jwalton** [comment]
[This article](https://remysharp.com/2015/10/26/using-travis-with-private-npm-deps) suggests generating your .npmrc file in travis-ci in the "before_install" step.  Most CIs offer some similar functionality - this makes it so your various users don't need to have NPM_TOKEN set - they just need to be logged in to NPM as a user that has permission to read packages.

Having the ability to generate a read-only token for the organization is a must, though - creating and managing an extra $7/month user just to deploy is very clunky.

**Comment by eduards** [comment]
I must agree and am wondering why this feature isn't available yet: Having the option of generating read-only token for both technical users such as CIs and for human users which should be only permitted to read the packages is absolutely essential. Hopefully this will be introduced soon...

**Comment by terinjokes** [comment]
Depending on how your CI initializes the environment, you can tell it to set the variables `npm_config_//registry.npmjs.org/:_authToken` and `npm_config__auth` and not have to templatize the file at all. Your shell probably won't let you set the former directly, but a program can set it then fork to a shell (and your CI might be setting up the environment this way already). 😄 

I note, that while these two configuration options are pretty vital, they aren't actually part of the [npm-config](https://docs.npmjs.com/misc/config) documentation. So, Caveat Emptor.

**Comment by ealves-pt** [comment]
I agree with @jwalton, an extra $7/month to have a user with readonly keys is very weird.

**Comment by sfarthin** [comment]
Is there any update on this? Is generating .npmrc in Travis CI still the recommended way?

**Comment by kaievns** [comment]
well this bug is generating the npm owners an extra $7/month, so they have no incentive to fix it, i suppose

**Comment by mathieug** [comment]
[Publishing what you mean to publish](http://blog.npmjs.org/post/165769683050/publishing-what-you-mean-to-publish) > **Using read-only tokens**
It seems we cannot create a read-only token for an organization but it's a first step.
